python/ray/tune/integration/kubernetes.py

```diff
@@ -0,0 +1,151 @@
+    ``KubernetesSyncClient`` instead.
+
+    KubernetesSyncer uses the default namespce ``ray``. You should
+    probably use ``NamespacedKubernetesSyncer`` to return a class
+    with a custom namespace instead.
+    """
+
+    _namespace = "ray"
+
+    def __init__(self, local_dir, remote_dir, sync_client=None):
+        self.local_ip = services.get_node_ip_address()
+        self.local_node = self._get_kubernetes_node_by_ip(self.local_ip)
+        self.worker_ip = None
+        self.worker_node = None
+
+        sync_client = sync_client or KubernetesSyncClient(
+            namespace=self.__class__._namespace)
+
+        super(NodeSyncer, self).__init__(local_dir, remote_dir, sync_client)
+
+    def set_worker_ip(self, worker_ip):
+        self.worker_ip = worker_ip
+        self.worker_node = self._get_kubernetes_node_by_ip(worker_ip)
+
+    def _get_kubernetes_node_by_ip(self, node_ip):
+        """Return node name by internal or external IP"""
+        kubernetes.config.load_incluster_config()
+        api = kubernetes.client.CoreV1Api()
+        pods = api.list_namespaced_pod(self._namespace)
+        for pod in pods.items:
+            if pod.status.host_ip == node_ip or \
+               pod.status.pod_ip == node_ip:
+                return pod.metadata.name
+
+        logger.error(
+            "Could not find Kubernetes node name for IP {}".format(node_ip))
+        return None
+
+    @property
+    def _remote_path(self):
+        return (self.worker_node, self._remote_dir)
+
+
+class KubernetesSyncClient(SyncClient):
+    """KubernetesSyncClient to be used by KubernetesSyncer.
+
+    This client takes care of executing the synchronization
+    commands for Kubernetes clients. In its ``sync_down`` and
+    ``sync_up`` commands, it expects tuples for the source
+    and target, respectively, for compatibility with the
+    KubernetesCommandRunner.
+
+    Args:
+        namespace (str): Namespace in which the pods live.
+        process_runner: How commands should be called.
+            Defaults to ``subprocess``.
+
+    """
+
+    def __init__(self, namespace, process_runner=subprocess):
+        self.namespace = namespace
+        self._process_runner = process_runner
+        self._command_runners = {}
+
+    def _create_command_runner(self, node_id):
+        """Create a command runner for one Kubernetes node"""
+        return KubernetesCommandRunner(
+            log_prefix="KubernetesSyncClient: {}:".format(node_id),
+            namespace=self.namespace,
+            node_id=node_id,
+            auth_config=None,
+            process_runner=self._process_runner)
+
+    def _get_command_runner(self, node_id):
+        """Create command runner if it doesn't exist"""
+        if node_id not in self._command_runners:
+            command_runner = self._create_command_runner(node_id)
+            self._command_runners[node_id] = command_runner
+        return self._command_runners[node_id]
```

Comment:
When do entries in this cache get cleaned up?

Reply:
Not seeing it anywhere - at a minimum I'd leave a comment for future work

Reply:
I'll think about this and address it later.

Reply:
yeah I think you end up with 1 command runner for each syncer, and one syncer for each trial

Reply:
It should be very cheap right now, as the command runners don't carry much logic and only call external programs. I left a note to address this if we proceed to use the syncer more (e.g. for log syncing or between-worker syncing).